pkg/simple/client/openpitrix/helmwrapper/helm_wrapper.go

```diff
@@ -470,142 +344,107 @@ func (c *helmWrapper) Install(chartName, chartData, values string) (err error) {
+	return install.Run(chart, values)
+}
 
-	if len(values) > 0 {
-		cmd.Args = append(cmd.Args, "--values", filepath.Join(c.Workspace(), "values.yaml"))
+func (c *helmWrapper) writeAction(chartName, chartData, values string, upgrade bool) error {
+	if klog.V(2) {
+		start := time.Now()
+		defer func() {
+			klog.V(2).Infof("run command end, namespace: %s, name: %s, upgrade: %t, elapsed: %v", c.Namespace, c.ReleaseName, upgrade, time.Now().Sub(start))
+		}()
 	}
 
-	if c.dryRun {
-		cmd.Args = append(cmd.Args, "--dry-run")
+	if err := c.ensureWorkspace(); err != nil {
+		return err
 	}
+	defer c.cleanup()
 
-	if c.kubeConfigPath() != "" {
-		cmd.Args = append(cmd.Args, "--kubeconfig", c.kubeConfigPath())
+	if err := c.createChart(chartName, chartData, values); err != nil {
+		return err
 	}
+	klog.V(8).Infof("namespace: %s, name: %s, chart values: %s", c.Namespace, c.ReleaseName, values)
 
-	// Post render, add annotations or labels to resources
-	if len(c.labels) > 0 || len(c.annotations) > 0 {
-		cmd.Args = append(cmd.Args, "--post-renderer", filepath.Join(c.Workspace(), postRenderExecFile))
+	chartRequested, err := loader.Load(c.chartPath())
+	if err != nil {
+		return err
 	}
-
-	if klog.V(8) {
-		// output debug info
-		cmd.Args = append(cmd.Args, "--debug")
+	valuePath := filepath.Join(c.Workspace(), "values.yaml")
+	helmValues, err := chartutil.ReadValuesFile(valuePath)
+	if err != nil {
+		return err
 	}
 
-	klog.V(4).Infof("run command: %s", cmd.String())
-	err = cmd.Run()
-
-	if err != nil {
-		klog.Errorf("namespace: %s, name: %s, run command: %s failed, stderr: %s, error: %v", c.Namespace, c.ReleaseName, cmd.String(), stderr, err)
-		// return the error of helm install
-		return errors.New("%s", stderr.String())
+	var rel *helmrelease.Release
+	if c.mock {
+		rel, err = c.mockRelease()
 	} else {
-		klog.V(2).Infof("namespace: %s, name: %s, run command success", c.Namespace, c.ReleaseName)
-		klog.V(8).Infof("namespace: %s, name: %s, run command success, stdout: %s", c.Namespace, c.ReleaseName, stdout)
+		if upgrade {
+			rel, err = c.helmUpgrade(chartRequested, helmValues.AsMap())
+		} else {
+			rel, err = c.helmInstall(chartRequested, helmValues.AsMap())
+		}
 	}
 
-	return
-}
-
-func (c *helmWrapper) Manifest() (manifest string, err error) {
-	if err = c.ensureWorkspace(); err != nil {
-		return "", err
+	if err != nil {
+		klog.Errorf("namespace: %s, name: %s,  error: %v", c.Namespace, c.ReleaseName, err)
+		return err
 	}
-	defer c.cleanup()
 
-	stdout := &bytes.Buffer{}
-	stderr := &bytes.Buffer{}
-	cmd := exec.Cmd{
-		Path: c.cmdPath,
-		Dir:  c.Workspace(),
-		Args: []string{
-			c.cmdPath,
-			"get",
-			"manifest",
-			c.ReleaseName,
-			"--namespace",
-			c.Namespace,
-		},
-		Stderr: stderr,
-		Stdout: stdout,
-	}
+	klog.V(2).Infof("namespace: %s, name: %s, run command success", c.Namespace, c.ReleaseName)
+	klog.V(8).Infof("namespace: %s, name: %s, run command success, release: %v", c.Namespace, c.ReleaseName, rel)
```

Comment:
`rel` is a struct, if we use `%v` to format it, the output is difficult to read.

Reply:
We should encode `rel` to string.

Reply:
yes,It's enough to output rel.Manifest here